serif_shared/src/commonMain/kotlin/xyz/room409/serif/serif_shared/MatrixClient.kt

```diff
@@ -0,0 +1,86 @@
+package xyz.room409.serif.serif_shared
+import kotlinx.coroutines.*
+import io.ktor.client.*
+import io.ktor.client.request.*
+import io.ktor.client.features.json.*
+import io.ktor.client.features.json.serializer.KotlinxSerializer
+import io.ktor.http.*
+import kotlinx.serialization.*
+import kotlinx.serialization.json.*
+
+@Serializable
+data class LoginRequest(val type: String, val identifier: LoginIdentifier, val password: String, val initial_device_display_name: String) {
+    constructor(username: String, password: String): this(type="m.login.password",
+                                                          identifier=LoginIdentifier(type="m.id.user",user=username),
+                                                          password=password,
+                                                          initial_device_display_name="Serif")
+}
+@Serializable
+data class LoginIdentifier(val type: String, val user: String)
+
+@Serializable
+data class LoginResponse(val access_token: String)
+
+@Serializable
+data class RoomMessage(val msgtype: String, val body: String) {
+    constructor(body: String): this(msgtype="m.text",body=body)
+}
+@Serializable
+data class EventIdResponse(val event_id: String)
+
+
+sealed class MatrixState {
+    val version: String
+        get() {
+            return "Serif Matrix client, pre-alpha on ${Platform().platform}"
+        }
+}
+class MatrixLogin(val login_message: String, val mclient: MatrixClient): MatrixState() {
+    constructor(): this(login_message="Please enter your username and password\n",
+                        mclient=MatrixClient())
```

Comment:
A single MatrixClient/MatrixSession is passed along through the states! It's thrown away now to simulate logging-out, since we have only implemented login.

Reply:
lgtm